dali/operators/geometry/transform_base_op.h

```diff
@@ -0,0 +1,189 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_OPERATORS_GEOMETRY_TRANSFORM_BASE_OP_H_
+#define DALI_OPERATORS_GEOMETRY_TRANSFORM_BASE_OP_H_
+
+#include <vector>
+#include "dali/core/format.h"
+#include "dali/core/geom/mat.h"
+#include "dali/core/static_switch.h"
+#include "dali/kernels/kernel_manager.h"
+#include "dali/pipeline/data/types.h"
+#include "dali/pipeline/operator/op_spec.h"
+#include "dali/pipeline/workspace/workspace.h"
+#include "dali/pipeline/operator/operator.h"
+
+#define TRANSFORM_INPUT_TYPES (float)
+
+namespace dali {
+
+template <int... values>
+using dims = std::integer_sequence<int, values...>;
+
+template <typename T, int ndim>
+using affine_mat_t = mat<ndim+1, ndim+1, T>;
+
+/**
+ * @brief Base CRTP class for affine transform generators.
+ * The matrix definition comes from the actual TransformImpl implementation.
+ * As with any CRTP-based system, any non-private method can be shadowed by the TransformImpl class.
+ */
+template <typename Backend, typename TransformImpl>
+class TransformBaseOp : public Operator<Backend> {
+ public:
+  explicit TransformBaseOp(const OpSpec &spec)
+      : Operator<Backend>(spec),
+        nsamples_(spec.GetArgument<int>("batch_size")) {
+  }
+
+  bool CanInferOutputs() const override { return true; }
+
+  TransformImpl &This() noexcept { return static_cast<TransformImpl&>(*this); }
+  const TransformImpl &This() const noexcept { return static_cast<const TransformImpl&>(*this); }
+
+ protected:
+  bool SetupImpl(std::vector<OutputDesc> &output_descs, const workspace_t<Backend> &ws) override {
+    has_input_ = ws.template NumInput() > 0;
+    if (has_input_) {
+      auto &input = ws.template InputRef<Backend>(0);
+      const auto &shape = input.shape();
+
+      DALI_ENFORCE(input.type().id() == dtype_,
+        make_string("Unexpected input data type. Expected ", dtype_, ", got: ", input.type().id()));
+
+      DALI_ENFORCE(shape.sample_dim() == 2 &&
+                   shape.size() > 0 &&
+                   shape[0][1] == (shape[0][0] + 1),
+        make_string(
+          "The input, if provided, is expected to be a 2D tensor with dimensions "
+          "(ndim, ndim+1) representing an affine transform. Got: ", shape));
+      nsamples_ = shape.num_samples();
+      for (int i = 0; i < nsamples_; i++) {
+        DALI_ENFORCE(shape[i] == shape[0],
+          make_string("Samples are expected to have the same dimensionality. Got: ", shape));
+      }
+    }
+
+    ndim_ = This().ndim(ws);
+
+    output_descs.resize(1);  // only one output
+    output_descs[0].type = TypeTable::GetTypeInfo(dtype_);
+    output_descs[0].shape = uniform_list_shape(nsamples_, {ndim_, ndim_+1});
+    return true;
+  }
+
+  template <typename T>
+  void RunImplTyped(workspace_t<Backend> &ws, dims<>) {
+    DALI_FAIL(make_string("Unsupported number of dimensions ", ndim_));  // NOLINT
```

Comment:
```suggestion
    DALI_FAIL(make_string("Unsupported number of dimensions ", ndim_));
```
leftover from VALUE_SWITCH, I guess...